Narration

jsarabia opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Bible-Translation-Tools/Orature/776)
<!-- Reviewable:end -->

mXaln committed:
Narration Demo

mXaln committed:
Record card menu text properties

mXaln committed:
Created Pcm classes and Pcm converter

mXaln committed:
undo debug code

mXaln committed:
undo debug code 2

mXaln committed:
undo debug code

mXaln committed:
undo debug code

mXaln committed:
Moved pcm conversion to AudioConverter

mXaln committed:
Tests for audio conversion

mXaln committed:
Refactored Mp3FileReader

mXaln committed:
Merge branch 'mm-pcm-audio' into mm-waveform-2.0

mXaln committed:
Merge branch 'mm-narration-demo-2' into mm-waveform-2.0

mXaln committed:
fixed unimplemented methods

mXaln committed:
Chapter split

mXaln committed:
Merge branch 'mm-pcm-audio' into mm-narration

mXaln committed:
Merge branch 'mm-split-chapter' into mm-narration

mXaln committed:
Changed narration listviews data types to generic;
Moved cells out of control package for developers to implement how they want

mXaln committed:
Merge branch 'mm-narration-demo-2' into mm-narration

mXaln committed:
Book title padding

mXaln committed:
Merge branch 'mm-narration-demo-2' into mm-narration

mXaln committed:
Floating card text format

mXaln committed:
Merge branch 'mm-narration-demo-2' into mm-narration

mXaln committed:
splitting chapter, loading chunks

mXaln committed:
object initialization changed

mXaln committed:
Merge branch 'mm-narration-demo-2' into mm-narration

mXaln committed:
Rendering waveform images for chunks

mXaln committed:
Refactored call factories

mXaln committed:
Merge branch 'mm-narration-demo-2' into mm-narration

# Conflicts:
#	jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationRecordListView.kt

mXaln committed:
Refactored cell factories

mXaln committed:
Check for null chapter audio file

mXaln committed:
Intractable waveform

mXaln committed:
Merge branch 'mm-split-chapter' into mm-narration

mXaln committed:
Merge branch 'mm-narration-demo-2' into mm-narration

# Conflicts:
#	jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationRecordItem.kt

mXaln committed:
Scrolling to corresponding text card when waveform clicked

mXaln committed:
Playback cursor over waveform

mXaln committed:
Added inverted waveform image

mXaln committed:
merged observables getting card images;
added some styling

mXaln committed:
Changed orientation of the recording list to the right

mXaln committed:
Use `use` to close OutputStream

mXaln committed:
Implement close function in AudioFileReader

mXaln committed:
Merge branch 'mm-pcm-audio' into mm-split-chapter

mXaln committed:
Created SplitAudioOnCues and ImportChunks classes

mXaln committed:
Merge branch 'mm-split-chapter' into mm-narration

mXaln committed:
Close reader

mXaln committed:
Merge branch 'mm-split-chapter' into mm-narration

mXaln committed:
Refactored chunk audio player

mXaln committed:
Moved callback functions to data class

mXaln committed:
rolled back the javafx version

jsarabia committed:
Merge branch 'Orature-2' into mm-split-chapter

mXaln committed:
Record a chunk again functionality

mXaln committed:
A working version of narration

mXaln committed:
waveform grow

mXaln committed:
Completed continuous waveform while recording

mXaln committed:
Merge branch 'Orature-2' into mm-split-chapter

mXaln committed:
Merge branch 'mm-split-chapter' into mm-narration

# Conflicts:
#	jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/demo/ui/components/NarrationRecordCell.kt
#	jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/demo/ui/components/NarrationTextCell.kt
#	jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/demo/ui/fragments/NarrationFragment.kt
#	jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationRecordItem.kt
#	jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationRecordListView.kt
#	jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationTextItem.kt
#	jvm/controls/src/main/resources/css/narration.css
#	jvm/controls/src/main/resources/css/theme/dark-theme.css
#	jvm/controls/src/main/resources/css/theme/light-theme.css
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/viewmodel/AudioPluginViewModel.kt

mXaln committed:
Refactored to remove dependency from PluginActions class

mXaln committed:
Merge branch 'mm-split-chapter' into mm-narration

# Conflicts:
#	common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/SplitAudioOnCues.kt

mXaln committed:
Update PluginActions.kt

mXaln committed:
Merge branch 'mm-split-chapter' into mm-narration

mXaln committed:
temp filename length should be more that 3 characters

mXaln committed:
Merge branch 'mm-split-chapter' into mm-narration

mXaln committed:
Refactored to use InsertTakeAudio class

mXaln committed:
Some refactoring;
Implemented chapter selection;
Opening chapter in verse marker app and external app

jsarabia committed:
Update jooq to 3.14.0

Updates jooq to 3.14.0, which supports Kotlin code gen,

As a result, types that were previously potentially null due to unannotated java types are nullable types. Asserting not null is done for now in order to keep consistent with the previous code.

DSL connections are not all implementing AutoCloseable (and instead, connections requiring closing are ClosableDSLConnection) thus the close/using functions were removed in jooq (and as they are no longer needed, they are removed from our code in this PR).

jsarabia committed:
Update jooq gradle plugin

jsarabia committed:
Update 3.14.0 to 3.14.16

3.15.0 and higher seems to cause issues in parsing the sql file to generate code, so 3.14.16 appears to be the highest version without introducing issues.

jsarabia committed:
Fix tests wrt kotlin codegen

jsarabia committed:
integrationTestImplementation instead of compile

jsarabia committed:
Ignore license plugin (incompatible with 7.0)

jsarabia committed:
Merge branch 'jbs-update-jooq' into jbs-update-gradle

jsarabia committed:
Update to gradle 7.0

jsarabia committed:
Update to gradle 7.6

mXaln committed:
Open chunk in an external app

mXaln committed:
Licence

mXaln committed:
Reset chapter

mXaln committed:
Correctly scroll to last item

mXaln committed:
Test fix

mXaln committed:
Using Ushort.MAX_VALUE

mXaln committed:
Merge branch 'Orature-2' into mm-narration

mXaln committed:
Undo/redo implementation

mXaln committed:
Some refactoring

mXaln committed:
Refactored NarrationHistory

mXaln committed:
abstract class RecordingRenderer

mXaln committed:
Create chunks from verses

mXaln committed:
Merge branch 'Orature-2' into mm-narration

# Conflicts:
#	jvm/workbookapp/src/main/resources/Messages_en.properties

jsarabia committed:
Begin NarrationView

jsarabia committed:
wip

jsarabia committed:
Merge branch 'Orature-2' into pair-narration

jsarabia committed:
Narration Header Layout

Refactors the layout for the Narration Header

jsarabia committed:
Refactor StickyVerse, add NarrationFooter

jsarabia committed:
Merge branch 'Orature-2' into jbs-update-gradle

mXaln committed:
Refactored NarrationTextCell and NarrationTextListView

mXaln committed:
Refactored InitialSelectedItemProperty to InitialSelectedVerseChangedEvent

jsarabia committed:
Use Chunk as backing data for NarrationTextList

Moves the narration text list to using Chunk rather than ChunkData

Moves the source of recording status out of the individual chunks as this is more of a global narration page status.

jsarabia committed:
Merge branch 'Orature-2' into pair-narration

jsarabia committed:
Merge branch 'Orature-2' into jbs-update-gradle

jsarabia committed:
Merge branch 'jbs-update-gradle' into pair-narration

jsarabia committed:
Mock Workbook Data for Demo App

jsarabia committed:
Init styles

jsarabia committed:
More workbook mocking

mXaln committed:
Do not call onDock for embedded views manually, because onDock will be called anyway

mXaln committed:
Created NarrationActions and NarrationHistory and their related tests

mXaln committed:
Ignoring undo/redo if stacks are empty

mXaln committed:
Fixed a bug with incorrect end position of the verse node

mXaln committed:
Added ResetAll Action

mXaln committed:
Changed Stack to ArrayDeque as an optimization

jsarabia committed:
Keep active verse in viewport on next

Keeps the active verse from getting moved slightly off screen when the next verse is selected.

mXaln committed:
Implemented basic record feature

mXaln committed:
Don't create new verse node onNext if state is pause

mXaln committed:
renaming

mXaln committed:
Undo/Redo functionality

mXaln committed:
Create temp file for narration;
Autoscroll to resumable item;

mXaln committed:
Load chapter details into the header and show availability of the file

mXaln committed:
Add ability to change chapters

mXaln committed:
use runLater instead of Thread class

mXaln committed:
Allow to open verse node in an external app

mXaln committed:
Allow to open chapter file in an external app / marker app

mXaln committed:
Merge branch 'mm-header-stuff' into mm-reset-on-edit

# Conflicts:
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationBody.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationMain.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationView.kt

mXaln committed:
Merge branch 'mm-chapter-external-app' into mm-reset-on-edit

# Conflicts:
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationView.kt

mXaln committed:
Update narration in-progress files after chapter was edited

mXaln committed:
Added javadocs and refactored "record again" feature

mXaln committed:
Add listener to the list of disposable listeners

mXaln committed:
refactored history saved file;
refactored last selected item;

mXaln committed:
Removed redundant event

mXaln committed:
Merge branch 'mm-open-versenode-externally' into mm-update-on-edit

# Conflicts:
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationBody.kt

mXaln committed:
Merge branch 'mm-header-stuff' into mm-update-on-edit

mXaln committed:
Merge branch 'mm-narration-temp-file' into mm-update-on-edit

# Conflicts:
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationBody.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationView.kt

mXaln committed:
Merge branch 'mm-chapter-external-app' into mm-update-on-edit

mXaln committed:
Merged with other branches and fixed conflicts

mXaln committed:
Renamed event

mXaln committed:
moved file name to constant

mXaln committed:
Merge branch 'mm-narration-temp-file' into mm-update-on-edit

# Conflicts:
#	common/src/main/kotlin/org/wycliffeassociates/otter/common/data/narration/NarrationHistory.kt

mXaln committed:
add helper function to get selected take for chapter

mXaln committed:
Merge branch 'mm-open-versenode-externally' into mm-narration-temp-file

# Conflicts:
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationBody.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationView.kt

mXaln committed:
resolved merge conflicts

mXaln committed:
Narration header chapter functionality (#761)

* Load chapter details into the header and show availability of the file

* Add ability to change chapters

* Add listener to the list of disposable listeners

mXaln committed:
Merge branch 'mm-narration-temp-file' into pair-narration

# Conflicts:
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationBody.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationMain.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationView.kt

mXaln committed:
Merge branch 'mm-chapter-external-app' into pair-narration

# Conflicts:
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationBody.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationHeader.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationMain.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationView.kt

mXaln committed:
resolved merge conflicts

mXaln committed:
Merge branch 'mm-update-on-edit' into pair-narration

# Conflicts:
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationBody.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationMain.kt
#	jvm/workbookapp/src/main/kotlin/org/wycliffeassociates/otter/jvm/workbookapp/ui/narration/NarrationView.kt

mXaln committed:
resolved merge conflicts

mXaln committed:
Created Narration class (#775)

* Created Narration class

* serialize and load from serialized text active verses

* fix tests

* more refactoring

* moved recorder and writer initialization to Narration class

* moved working files to ChapterRepresentation class;
refactored Narration load

* Refactored ChapterRepresentation class

* Remove SplitAudioOnCues dependency from view model

* Moved player to narration, removed audio utils dependency from the view model

* refactored injecting player and recorder

* more refactoring

* code reorganization

* updated dagger to 2.47;
created NarrationFactory to inject Narration

* added narration function to get recorder audio stream

* added ability to just reload narration from chapter file

* simplified updating narration with chapter file

* renamed function onChapterLoad

* more renaming

* changed how verse markers action works

* replace verse

* add api to get working audio in narration

jsarabia committed:
Orature narration merge (#777)

* Search bar - UI Control (#741)

* Common styles for text (#743)

defines common style classes for global use

* Update to Gradle 7 (#746)

* Update jooq to 3.14.0

Updates jooq to 3.14.0, which supports Kotlin code gen,

As a result, types that were previously potentially null due to unannotated java types are nullable types. Asserting not null is done for now in order to keep consistent with the previous code.

DSL connections are not all implementing AutoCloseable (and instead, connections requiring closing are ClosableDSLConnection) thus the close/using functions were removed in jooq (and as they are no longer needed, they are removed from our code in this PR).

* Update jooq gradle plugin

* Update 3.14.0 to 3.14.16

3.15.0 and higher seems to cause issues in parsing the sql file to generate code, so 3.14.16 appears to be the highest version without introducing issues.

* Fix tests wrt kotlin codegen

* integrationTestImplementation instead of compile

* Ignore license plugin (incompatible with 7.0)

* Update to gradle 7.0

* Update to gradle 7.6

* Enable license plugin

* Update GPL License (#749)

* Add Licenses

* Update licenses to include 2023

* Language Table (#748)

* change scrollbar style

* wip

* refactor

* refactor table value factory

* styling & localization

* set up table in demo

* update styling

* disable mouse effect on empty rows

* replace callback with event

* remove header

* fix

* discard new home page

* refactor

* remove header from table demo

* refactor css

* refactor popup to class

* rename data class

* header wip

* Merge branch 'Orature-2' into aw-book-table-header

* refactor search bar component

* clean up after merge

* remove unused

* bind prompt text prop

* h3, h4, h5 and icon default

* rename class

* rename

* table

* add action event for language selection

* normal text label

* Merge branch 'Orature-2' into aw-language-table

* remove unused styles

* Book Info model repository (#747)

* add slug & styles

* workbook info repository

* refactor repo

* rename info prop

* move to workbook repo

* rename to ProjectInfo

* Merge branch 'Orature-2' into aw-homepage-books-model

* update repo and rename to WorkbookDescriptor

* Home page: Translation cards (#744)

* change scrollbar style

* set up table in demo

* disable mouse effect on empty rows

* replace callback with event

* refactor popup to class

* h3, h4, h5 and icon default

* create translation card - new, active, default card

* separate demo view

* rename create translation

* use control & skin

* remove unused card

* use action property

* relocate ProjectMode enum

* Create Translation Wizard (#745)

* change scrollbar style

* wip

* refactor

* refactor table value factory

* styling & localization

* set up table in demo

* update styling

* disable mouse effect on empty rows

* replace callback with event

* remove header

* fix

* discard new home page

* refactor

* remove header from table demo

* refactor css

* refactor popup to class

* rename data class

* header wip

* refactor search bar component

* clean up after merge

* remove unused

* bind prompt text prop

* h3, h4, h5 and icon default

* create translation card

new, active, default card

* refactor

* rename class

* rename

* grid

* remove grid

* localized text

* action property

* separate demo view

* create demo view

* update style class

* doc

* text style

* add TODO

* rename create translation

* use control & skin

* remove unused card

* use action property

* use property for action

* Workbook descriptor schema (#752)

* dao and sql

* always insert workbook descriptor when deriving a project

* Home Page 2 (#753)

* dao and sql

* repo

* compile error

* workbook type table

* refactor

* derive all books when creating translation

* fix insertion order

* use entity id for wbDescriptor

* remove getter from old workbook repo

* comment

* use new API from rc lib

* documentation

* clean up UI after adding viewmodels

* touch up and refactor

* always insert workbook descriptor when deriving a project

* icon style

* handle toggling between fragments

* rename card model

* fix transaction in transaction SQL error

* open workbook

* clean up

* temporary fix use case

* change workbook descriptor constructor vals

* display available source languages only

* remove hashCode override

* remove equals override

* style active option button

* remove hashcode & equals in card model

* add breadcrumb to home

* prevent arranging columns

* Add anthology to workbook table (#755)

* dao and sql

* repo

* compile error

* workbook type table

* refactor

* derive all books when creating translation

* fix insertion order

* use entity id for wbDescriptor

* remove getter from old workbook repo

* comment

* use new API from rc lib

* documentation

* clean up UI after adding viewmodels

* touch up and refactor

* always insert workbook descriptor when deriving a project

* icon style

* handle toggling between fragments

* rename card model

* fix transaction in transaction SQL error

* open workbook

* clean up

* temporary fix use case

* change workbook descriptor constructor vals

* display available source languages only

* add anthology

* remove hashCode override

* remove equals override

* style active option button

* remove "other"

* remove hashcode & equals in card model

* add breadcrumb to home

* prevent arranging columns

* prevent arranging column

* Search filter (#756)

* search then sort

* Handle table sort binding collisions (#757)

* refactor sort table with sort order by Bible order

* Home page accessibility (#758)



* fix bug focus on first item

* do not focus on active card

* Resize element width in Home page (#759)



* do not focus on active card

* progress bar & left pane

* scrollpane for project groups (left pane)

* book code column title

* Delete project group (#754)

* delete a single project will reset and re-insert it to the project group

* Remove failed & unnecessary test (#768)

* temporarily ignore failed test

* delete test

* Remove import (#769)

* temporarily ignore failed test

* delete test

* import

* Project Import dialog (#763)


* relocate ui model

* refresh home page after import

to avoid inconsistency between database and ui

* avoid checking docked - event subscription won't reach when undocked

* Project Export Dialog (#765)


* relocate ui model

* checkbox selection with accessibility

* update styling

* rename import-export-dialog.css

* remove debug code

* export view model

* apply filter to exporter api

* select all checkbox

* prevent UI input while browsing files

* remove unused props

* refresh home page after import

to avoid inconsistency between database and ui

* avoid checking docked

event subscription won't reach when undocked

* merge clean up

* rename class ChapterDescriptor

* relocate chapter data class

* remove debug code

* Notifications of Import/Export (#771)

* wip - working import

* ui & handler for import conflict

* relocate css import

* css

* message

* Revert "css"

This reverts commit f538b0329fd1e6ef3a7818fd181659518acb2622.

* Merge branch 'Orature-2' into aw-import-ui

* basic UI

* relocate ui model

* checkbox selection with accessibility

* update styling

* rename css file

* remove debug code

* export view model

* minor css

* apply filter to exporter api

* select all checkbox

* prevent UI input while browsing files

* remove unused props

* refresh home page after import

to avoid inconsistency between database and ui

* Merge branch 'aw-import-ui' into aw-export-ui

* avoid checking docked

event subscription won't reach when undocked

* Merge branch 'aw-import-ui' into aw-export-ui

* merge clean up

* line typo

* remove debug code

* wip

* resolve different snackbar permutations

* notification message event

* export error notification

* notification for import ongoing project

* notification for source import

* open book callback

* documentation & clean up

* comment

* update test

* Merge branch 'Orature-2' into aw-snackbar-notification

* errors after merge

* refactor snackbar ui

* Delete WorkbookPageViewModelTest.kt

* update api doc

* refactor

* fix bug delete project

* clean up

* Import/Export progress dialog (reopened #772) (#774)

* wip - working import

* ui & handler for import conflict

* relocate css import

* css

* message

* Revert "css"

This reverts commit f538b0329fd1e6ef3a7818fd181659518acb2622.

* basic UI

* relocate ui model

* checkbox selection with accessibility

* update styling

* rename css file

* remove debug code

* export view model

* minor css

* apply filter to exporter api

* select all checkbox

* prevent UI input while browsing files

* remove unused props

* refresh home page after import

to avoid inconsistency between database and ui

* avoid checking docked

event subscription won't reach when undocked

* merge clean up

* line typo

* remove debug code

* wip

* resolve different snackbar permutations

* notification message event

* export error notification

* notification for import ongoing project

* notification for source import

* open book callback

* documentation & clean up

* comment

* update test

* errors after merge

* refactor snackbar ui

* Delete WorkbookPageViewModelTest.kt

* export progress update

* css

* add message to export progress

* update api doc

* refactor

* fix bug delete project

* refactor

* fix bug chapters not selected

* clean up

* refactor snackbar init

* wip

* progress dialog for import

* message for import source of resumable project

* fix test error

* doFinally instead of doOnComplete

* merge conflict

* progress message for new source import

* Quick backup for project export (#773)

* New Crowdin updates (#736)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

* New translations Messages_en.properties (French)

* New translations Messages_en.properties (Arabic)

* New translations Messages_en.properties (Chinese Simplified)

* New translations Messages_en.properties (Vietnamese)

* New translations Messages_en.properties (Portuguese, Brazilian)

* New translations Messages_en.properties (Telugu)

* New translations Messages_en.properties (Burmese)

* New translations Messages_en.properties (Swahili)

* New translations Messages_en.properties (Spanish, Latin America)

---------

Co-authored-by: Tony Tran <34975907+AnonymousWalker@users.noreply.github.com>
Co-authored-by: crowdin-wa-btt <95763661+crowdin-wa-btt@users.noreply.github.com>

jsarabia committed:
Merge branch 'Orature-2' into pair-narration

jsarabia committed:
Update AppDependencyGraph.kt

jsarabia committed:
cleanup pair narration (#779)

* uncomment license plugin code

* Move Verse Node

VerseNode isn't really a primitive to be used of all of Orature, but just a data type used in the Narration package.

* Remove "Saved in db" from Take

Removed savedInDb- the Take object should have no coupling to a persistence concept.

* Remove savedInDb calls

* Revert changes to the ActiveRecordingRenderer

* Revert Renderers

* Remove old/unused narration components

jsarabia committed:
Merge branch 'Orature-2' into pair-narration

jsarabia committed:
Use OratureAudioFile

jsarabia committed:
Update dependencies.gradle

jsarabia committed:
Revert "Update dependencies.gradle"

This reverts commit 2c47cebf5239e2d1d83805875234129f96846969.

jsarabia committed:
Add mockk

jsarabia committed:
Comment out tests that broke from upgrading kotlin (needs a different mocking framework, or refactoring)

jsarabia committed:
Update VerseMarkerViewModel.kt

mXaln committed:
Return modifying verse end position

mXaln committed:
Implemented AudioFileReader in ChapterRepresentation (#778)

* Implemented AudioFileReader in ChapterRepresentation

* Optimized chapter representation

* Fixed merge conflicts

* Added tests for ChapterRepresentation as file reader

* Refactored reading pcm stream

* Removed duplicates

jsarabia committed:
Connect Narration (#790)

Connects Narration to the HomePage

jsarabia committed:
Refactor narration viewmodels (#794)

* Move event subscriptions to top level viewmodel

Much of the major narration events for the entire page were handled by the AudioWorkspace, which, should simply be responsible for the visualization and marker interactions. Many such events effect the page as a whole, and thus, need to have data propagated downward to the other ViewModels.

jsarabia committed:
Refactor versenode (#800)

* Allocate total verses, keep track of "placed" markers

Rather than adding verses, all markers should be allocated (due to needing the labels to be sourced elsewhere).

As a result, we need to keep track of whether markers are "placed" or not.

Active verses can just be total verses filtered on whether they are placed.

Synchronization needed as a result of multiple thread access.

AnonymousWalker commented on common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/AudioFileUtils.kt:
```diff
@@ -0,0 +1,39 @@
+        audio.reader(start, end).use { reader ->
+            pcmWriter.use { writer ->
+                reader.open()
+                val buffer = ByteArray(10240)
```
const val BUFFER_SIZE?

AnonymousWalker commented on common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/ChapterRepresentation.kt:
```diff
@@ -0,0 +1,216 @@
+private const val ACTIVE_VERSES_FILE_NAME = "active_verses.json"
+private const val CHAPTER_NARRATION_FILE_NAME = "chapter_narration.pcm"
+
+internal class ChapterRepresentation(
```
Could we add documentation to describe this class? It'd be helpful to grab the general idea when encountering this abstract name

AnonymousWalker commented on common/src/main/kotlin/org/wycliffeassociates/otter/common/domain/narration/SplitAudioOnCues.kt:
```diff
@@ -0,0 +1,97 @@
+        }
+    }
+
+    fun execute(file: File, cues: List<VerseMarker>): Single<VerseSegments> {
```
is this supposed to be executeAsync, too?

AnonymousWalker commented on jvm/controls/src/main/kotlin/org/wycliffeassociates/otter/jvm/controls/narration/NarrationTextListView.kt:
```diff
@@ -18,76 +18,71 @@
 import javafx.scene.control.ScrollBar
-import org.wycliffeassociates.otter.jvm.utils.findChildren
-import org.wycliffeassociates.otter.jvm.utils.virtualFlow
+import org.wycliffeassociates.otter.jvm.utils.*
```
\*

jsarabia committed:
Use VerseMarker as a public API, VerseNode internal to ChapterRepresentation (#805)

* Move event subscriptions to top level viewmodel

Much of the major narration events for the entire page were handled by the AudioWorkspace, which, should simply be responsible for the visualization and marker interactions. Many such events effect the page as a whole, and thus, need to have data propagated downward to the other ViewModels.

* Allocate total verses, keep track of "placed" markers

Rather than adding verses, all markers should be allocated (due to needing the labels to be sourced elsewhere).
As a result, we need to keep track of whether markers are "placed" or not.
Active verses can just be total verses filtered on whether they are placed.
Synchronization needed as a result of multiple thread access.

jsarabia committed:
Merge branch 'Orature-2' into pair-narration

jsarabia committed:
VerseNode Ranges (#817)

* Replace end-start calculation with a function

* Refactor Verse Node to contain a range of sectors, not just a start and end

As VerseMarkers can be moved around, bad sectors can begin to arise in the Scratch Audio File and Verses can span across these bad sectors. Rather than one continuous range, this then needs to be accounted for by a list of sectors.

* Introduce ReaderProvider

Rather than get an AudioFileReader from a file, this allows getting the reader other ways, which will be necessary to allow for loading ChapterRepresentation connections into the Audio system

* Update Jackson

Updates the jackson library in order to allow for serialization of IntRange

Adds org.json as a dependency due to the github bug report class- presumably this was previously provided via jackson but now needs it explicitly?

* Fix VerseMarkerAction

Fixes the verse marker action to either take or give frames to adjacent verse nodes.

jsarabia committed:
Add waveform renderer (#821)

* WaveformDrawables

* Get location and duration from Narration, include active recording

Location in regards to narration and rendering should be with respect to the actual location which is a combination of the player position as well as what may be actively recorded from the microphone.

* Don't seek back to 0 at the end of AudioPlayback

Rather than jumping back to the beginning of the audio file, we should stay where we are.

jsarabia committed:
Fix frameSizeBytes

jsarabia committed:
Change pcm reader (#824)

Changes the PCM reader to play one verse at a time, and iterate over sectors

* Try reading a verse at a time

* Lock playback of recordings

jsarabia committed:
Bugfix: don't coerceIn a position out of bounds

frames and sectors can be out of sequential order and unpredictable, if the position is not in the node, move the read head to the first frame of the node.

jsarabia committed:
Fix CSS for Pause button

Adds the "active" pseudoclass for the play/pause toggle.

jsarabia committed:
Various narration bugfixes (#828)

* Loading modal in Home page (#808)

* loading dialog & handle post-load

* Merge branch 'Orature-2' into aw-loading-home-page

* Add custom scrollbar skin in Home page (#811)

* chunk items

* refactor

* refactor

* wip

* clean up

* chunking steps

* accessibility + stylings

* disable unreachable steps

* show/hide completed steps

* show/hide completed icon

* scrollable side pane

* completed steps graphic

* toggle collapsed pane

* stretch grid items horizontally

* clean up

* const for grid cols

* use const

* use variable for status completed

* update hide completed logic

* custom scrollbar graphic & styling

* style for scroll bar arrows

* refactor

* refactor

* comment

* Merge branch 'Orature-2' into aw-custom-scrollbar

* adjust size

* Merge branch 'Orature-2' into aw-custom-scrollbar

* add scrollbar to scroll pane in home

* animate when selecting project (#812)

* Chapter selector grid (#793)

* chunk items

* refactor

* refactor

* wip

* clean up

* chunking steps

* accessibility + stylings

* disable unreachable steps

* show/hide completed steps

* show/hide completed icon

* scrollable side pane

* completed steps graphic

* toggle collapsed pane

* stretch grid items horizontally

* clean up

* const for grid cols

* create chapter grid selector

* use const

* Merge branch 'aw-new-chunking' into aw-chapter-selector

* padding

* Merge branch 'Orature-2' into aw-chapter-selector

* item property for grid

* use weak listener

* Revert "use weak listener"

This reverts commit 7837872b3e4ce944d9a96d593bbe955f825db8c1.

* Revert "item property for grid"

This reverts commit f10bef01e91fe579cc419285d93659bce09d2c5f.

* Undo delete project (#814)

* loading dialog & handle post-load

* Merge branch 'Orature-2' into aw-loading-home-page

* animate when selecting project

* undo delete with cancellable task

* Merge remote-tracking branch 'origin/aw-loading-home-page' into aw-undo-delete-project

* handle undo with UI only

* Merge branch 'Orature-2' into aw-undo-delete-project

* Stores contributor information in derived rc (#815)

* persist contributors in der/ container

* use observable for async load/save

* Source text pane (#816)

* chunk items

* refactor

* refactor

* wip

* clean up

* chunking steps

* accessibility + stylings

* disable unreachable steps

* show/hide completed steps

* show/hide completed icon

* scrollable side pane

* completed steps graphic

* toggle collapsed pane

* stretch grid items horizontally

* clean up

* const for grid cols

* use const

* use variable for status completed

* update hide completed logic

* custom scrollbar graphic & styling

* style for scroll bar arrows

* refactor

* refactor

* comment

* wip

* source text list refactor

* new demo view

* license example

* Merge branch 'Orature-2' into aw-source-text-pane

* Merge branch 'Orature-2' into aw-source-text-pane

* Add scrollbar visual fix

* Stop seeking around after moving markers

* Fix Scrollbar

Binding seek to the scrollbar value will cause various issues, disable it while playing and incorporate updating it into the render loop.

* Scroll waveform

* Fix active recording while navigating to next verses

* Don't allow markers to move on top of one another.

* Fix seek to map relativeToAbsolute

* Off by 1 in RecordAgainAction

Corrects an off by 1 error by starting record again at the next frame to begin when recording at the end of the file.

* Fix some off by ones

* Guard against empty lists for the case of Chapter Resets

* Public API for framePosition needs to be in relative units

* Remove start/end scratch frame

* Remove start/end from VerseNode

* Guard AudioReaderDrawable from infinite loop

---------

Co-authored-by: Tony <34975907+AnonymousWalker@users.noreply.github.com>

jsarabia committed:
More narration fixes (#829)

* Remove start/end from VerseNode

* Guard AudioReaderDrawable from infinite loop

* Fix launching verse in OcenAudio

jsarabia committed:
Narration chapter take (#831)

* Create Chapter takes when chapters are potentially finished.

* Bind marker label width to text length

jsarabia committed:
More narration fixes (#832)

* Loading modal in Home page (#808)

* loading dialog & handle post-load

* Merge branch 'Orature-2' into aw-loading-home-page

* Add custom scrollbar skin in Home page (#811)

* chunk items

* refactor

* refactor

* wip

* clean up

* chunking steps

* accessibility + stylings

* disable unreachable steps

* show/hide completed steps

* show/hide completed icon

* scrollable side pane

* completed steps graphic

* toggle collapsed pane

* stretch grid items horizontally

* clean up

* const for grid cols

* use const

* use variable for status completed

* update hide completed logic

* custom scrollbar graphic & styling

* style for scroll bar arrows

* refactor

* refactor

* comment

* Merge branch 'Orature-2' into aw-custom-scrollbar

* adjust size

* Merge branch 'Orature-2' into aw-custom-scrollbar

* add scrollbar to scroll pane in home

* animate when selecting project (#812)

* Chapter selector grid (#793)

* chunk items

* refactor

* refactor

* wip

* clean up

* chunking steps

* accessibility + stylings

* disable unreachable steps

* show/hide completed steps

* show/hide completed icon

* scrollable side pane

* completed steps graphic

* toggle collapsed pane

* stretch grid items horizontally

* clean up

* const for grid cols

* create chapter grid selector

* use const

* Merge branch 'aw-new-chunking' into aw-chapter-selector

* padding

* Merge branch 'Orature-2' into aw-chapter-selector

* item property for grid

* use weak listener

* Revert "use weak listener"

This reverts commit 7837872b3e4ce944d9a96d593bbe955f825db8c1.

* Revert "item property for grid"

This reverts commit f10bef01e91fe579cc419285d93659bce09d2c5f.

* Undo delete project (#814)

* loading dialog & handle post-load

* Merge branch 'Orature-2' into aw-loading-home-page

* animate when selecting project

* undo delete with cancellable task

* Merge remote-tracking branch 'origin/aw-loading-home-page' into aw-undo-delete-project

* handle undo with UI only

* Merge branch 'Orature-2' into aw-undo-delete-project

* Stores contributor information in derived rc (#815)

* persist contributors in der/ container

* use observable for async load/save

* Source text pane (#816)

* chunk items

* refactor

* refactor

* wip

* clean up

* chunking steps

* accessibility + stylings

* disable unreachable steps

* show/hide completed steps

* show/hide completed icon

* scrollable side pane

* completed steps graphic

* toggle collapsed pane

* stretch grid items horizontally

* clean up

* const for grid cols

* use const

* use variable for status completed

* update hide completed logic

* custom scrollbar graphic & styling

* style for scroll bar arrows

* refactor

* refactor

* comment

* wip

* source text list refactor

* new demo view

* license example

* Merge branch 'Orature-2' into aw-source-text-pane

* Merge branch 'Orature-2' into aw-source-text-pane

* Add scrollbar visual fix

* Stop seeking around after moving markers

* Fix Scrollbar

Binding seek to the scrollbar value will cause various issues, disable it while playing and incorporate updating it into the render loop.

* Scroll waveform

* Fix active recording while navigating to next verses

* Don't allow markers to move on top of one another.

* Fix seek to map relativeToAbsolute

* Off by 1 in RecordAgainAction

Corrects an off by 1 error by starting record again at the next frame to begin when recording at the end of the file.

* Fix some off by ones

* Guard against empty lists for the case of Chapter Resets

* Public API for framePosition needs to be in relative units

* Remove start/end scratch frame

* Remove start/end from VerseNode

* Guard AudioReaderDrawable from infinite loop

* Fix launching verse in OcenAudio

darrellcolehill committed:
Dh pair narration cherry picked test (#833)

* Removed old test files and added most recent

* Resolved internal access error

* Resolved internal access and refactor

Creates test folders based on the testDataRootFilePath. The folders are needed to store mock audio data for testing

* Resolved error I am getting on Mac

* Removed unnecessary comments

* Added test for editVerseAction

* Updated working audio ends

Made working audio length grow when the number of test verses grows

* Revert change to ZipEntryTreeBuilder

---------

Co-authored-by: Joe <jsarabia1247@gmail.com>

darrellcolehill committed:
Dh pair narration cherry picked test (#834)

* Removed old test files and added most recent

* Resolved internal access error

* Resolved internal access and refactor

Creates test folders based on the testDataRootFilePath. The folders are needed to store mock audio data for testing

* Resolved error I am getting on Mac

* Removed unnecessary comments

* Added test for editVerseAction

* Updated working audio ends

Made working audio length grow when the number of test verses grows

* Notes and comments

Added clarification as to why EditVerseTest will fail, and added note to ask if we need to throw exceptions or not

* Refactored takeFramesFromStart/end test

Went ahead and assumed that we want to throw the exception

* Refactored and updated test

Refactored how test workingAudio is generated and updated test for relativeToAbsolute

* Potential partial workaround for hardcoding path

* Off by one fix

* Potential cross-platform fix for making test audio files

* Cleans up generated test audio files

* Initial Narration test

* Removed changes that are not relevant for upcoming PR

* Removed unnecessary test

jsarabia committed:
add missing imports

darrellcolehill committed:
Dh verse playback bug fix (#837)

* Bug fix for relativeToAbsolute

* Additional unit test for relativeToAbsolute

Addresses found bug

* Fixed spacing

jsarabia committed:
Narration teleprompter UI (#838)

* WIP

* WIP

* text fixes

* Play/Pause/Record index passed down

* Disable play button when lacking a recording

* Update TeleprompterView.kt

* Fix state of item after current recording

* Update NarrationViewModel.kt

* Add save recording

* Playback of verses

* More events

* Model each Narration state as an enum

* Disable recording when another verse is recording or playing

* Debounce teleprompter actions
